Add Aqara FP1E presence sensor v2 quirk

jeverley opened the pull request:
## Proposed change
<!--
  Explain your proposed change below.
-->
This quirk adds support for the Aqara FP1E presence sensor, exposing its motion and occupancy sensors in addition to its  configuration parameters.

Fixes https://github.com/zigpy/zha-device-handlers/issues/3294 and is written using the quirks v2 format.

![image](https://github.com/user-attachments/assets/07e1b493-89c8-4f5a-92f0-ad40cfbb2ab0)



## Additional information
<!--
  Please include any additional information that is important to this PR.
  For example, if this PR is a potentially breaking change, mention that here.
  If this PR requires other PRs to be merged in HA Core or other projects, mention that.
  Lastly, if this PR fixes a specific issue, please include "Fixes #xxxx".
-->

There are two issues that we'd ideally want to fix before merging this PR.

- https://github.com/zigpy/zha/issues/290
   - _Currently unable to specify the unit for the distance sensors_
   - **PR raised:** https://github.com/zigpy/zha/pull/307
- https://github.com/zigpy/zha/issues/291
   - _I'd like to make the _Restart device_ button a diagnostic rather than config function_
   - **PR raised:** https://github.com/zigpy/zha/pull/308

## Checklist
<!--
  Put an 'x' in all boxes that apply.
  Note: You do not need to tick all boxes before creating a PR.
-->

- [x] The changes are tested and work correctly
- [x] `pre-commit` checks pass / the code has been formatted using Black
- [ ] Tests have been added to verify that the new code works

jeverley committed:
Add v2 format quirk for Aqara FP1E presence sensor

Addresses https://github.com/zigpy/zha-device-handlers/issues/3294

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

jeverley committed:
Resolve mypy AttributeDefs error

jeverley committed:
Remove unused logger

jeverley commented:
Hi @dmulcahey, do you have any feedback on the current implementation with regards to quirk v2 conventions?

TheJulianJES commented on zhaquirks/xiaomi/aqara/motion_agl1.py:
```diff
@@ -0,0 +1,179 @@
+
+    def _update_attribute(self, attrid: int, value: Any) -> None:
+        super()._update_attribute(attrid, value)
+        if attrid == OCCUPANCY_ATTR_ID:
```
This would be more consistent with our other (new) quirks:
```suggestion
        if attrid == self.AttributeDefs.occupancy.id:
```
(or: `OppleCluster.AttributeDefs.occupancy.id`)

jeverley replied:
Thanks for the feedback @TheJulianJES , updated with https://github.com/zigpy/zha-device-handlers/pull/3521/commits/0636e87c22e7c247ad57ac9c690953e511745f72 to follow this recommendation.

TheJulianJES commented on zhaquirks/xiaomi/aqara/motion_agl1.py:
```diff
@@ -0,0 +1,179 @@
+                OccupancySensing.AttributeDefs.occupancy.id,
+                OccupancySensing.Occupancy(value),
+            )
+        elif attrid == MOTION_ATTR_ID:
```
and same here:
```suggestion
        elif attrid == self.AttributeDefs.motion.id:
```

TheJulianJES commented on zhaquirks/xiaomi/aqara/motion_agl1.py:
```diff
@@ -0,0 +1,179 @@
+(
+    QuirkBuilder("aqara", "lumi.sensor_occupy.agl1")
+    .friendly_name(model="Presence Sensor FP1E", manufacturer="Aqara")
+    .adds(DeviceTemperature)
+    .adds(OccupancySensing)
+    .adds(
+        IasZone,
+        constant_attributes={
+            IasZone.AttributeDefs.zone_type: IasZone.ZoneType.Motion_Sensor
+        },
+    )
```
If these clusters are completely virtual/fake (not present on the device), we'd probably want them to be a `LocalDataCluster`. E.g.:

```python
class IasZoneLocal(LocalDataCluster, IasZone):
    """Virtual cluster for IasZone."""

    # required to make sure ZHA creates sensors when initially pairing
    _VALID_ATTRIBUTES = {IasZone.AttributeDefs.zone_status.id}
```

`_VALID_ATTRIBUTES` is probably required to make sure that ZHA creates the sensors when the device is initially paired (so completely removed from ZHA before pairing again), even when there's no initial data in the zigpy DB (which there is not, as the attribute cannot be read on pairing).

TheJulianJES replied:
It might be a good idea for us to provide a shorthand v2 quirk method for this. But that does't exist yet..

jeverley replied:
> It might be a good idea for us to provide a shorthand v2 quirk method for this. But that does't exist yet..

Yes I think that would be a great addition, in the meantime I've updated the quirk to use Local clusters for motion and occupancy.

TheJulianJES commented on zhaquirks/xiaomi/aqara/motion_agl1.py:
```diff
@@ -0,0 +1,179 @@
+MOTION_SENSITIVITY_ATTR_ID = 0x010C  # UINT8 The configurable detection sensitivity (0x01 = low, 0x02 = medium, 0x03 = high, default 0x03 = High)
+OCCUPANCY_ATTR_ID = 0x0142  # UINT8 Occupancy detected 0x0 = absence, 0x01 = presence
+RESET_NO_PRESENCE_STATUS_ATTR_ID = 0x0157  # UINT8 Trigger AI spatial learning
+RESTART_DEVICE_ATTR_ID = 0x00E8  # BOOL Trigger device restart
```
We can probably remove these constants (define them directly in the `ZCLAttributeDef`) and move the comments above the respective `ZCLAttributeDef`.

jeverley replied:
Agreed that makes sense, the ZCLAttributeDefs allow a much more consistent approach.
I've addressed this in https://github.com/zigpy/zha-device-handlers/pull/3521/commits/0636e87c22e7c247ad57ac9c690953e511745f72.

jeverley committed:
Local reporting clusters and AttributeDefs changes

Move attribute IDs into the AttributeDefs and ensure that OccupancySensing and IasZone are correctly handled as local virtual clusters.

jeverley committed:
Address codespell error

jeverley committed:
Use types to document OppleCluster values

jeverley committed:
Restore docstrings on types

jeverley committed:
Post zigpy bug fix merges to unit and entity_type

UnitOfLength.METERS for distance entities (https://github.com/zigpy/zha/pull/307)

EntityType.DIAGNOSTIC for restart device button (https://github.com/zigpy/zha/pull/308)

jeverley committed:
Merge branch 'zigpy:dev' into lumi.sensor_occupy.agl1

TheJulianJES commented on zhaquirks/xiaomi/aqara/motion_agl1.py:
```diff
@@ -0,0 +1,194 @@
+    .adds(OccupancySensingLocal)
+    .adds(
+        IasZoneLocal,
+        constant_attributes={
+            IasZone.AttributeDefs.zone_type: IasZone.ZoneType.Motion_Sensor
+        },
```
I think I remember someone complaining about `constant_attributes` defined via quirks v2 not always working as expected. And I might see why: zigpy adds the attribute name as the key, not the attribute id. This might work in some places, but is unexpected in others. Zigpy code for reference: [zigpy/quirks/v2/__init__.py#L178-L182](https://github.com/zigpy/zigpy/blob/8e9f5163ef93ff71fa0ce4992076d8bc248eac44/zigpy/quirks/v2/__init__.py#L178-L182)

Until this is fixed in zigpy, let's remove this here and add it in the cluster class above via the attribute id in `_CONSTANT_ATTRIBUTES`. We're defining `_VALID_ATTRIBUTES` there anyway, so they're also together then.
We'll want to update this again when we add valid attributes as a parameters here for quirks v2 (and fix the constant attributes issue).
```suggestion
```

jeverley replied:
Good catch, thanks for making that change.

TheJulianJES commented on zhaquirks/xiaomi/aqara/motion_agl1.py:
```diff
@@ -0,0 +1,194 @@
+
+class IasZoneLocal(LocalDataCluster, IasZone):
+    """Virtual cluster for IasZone."""
+
```
So let's add the zone type as a constant attribute here, per the previous comment.
```suggestion

    _CONSTANT_ATTRIBUTES = {IasZone.AttributeDefs.zone_type.id: IasZone.ZoneType.Motion_Sensor}
```

TheJulianJES reviewed (commented):
This looks very good.
I'll make a small change regarding the `constant_attributes`. There's [more info ](https://github.com/zigpy/zha-device-handlers/pull/3521#discussion_r1857687985)in the two comments. Otherwise, this should be good to go.

TheJulianJES committed:
Use old-style `_CONSTANT_ATTRIBUTES`

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

TheJulianJES commented on zhaquirks/xiaomi/aqara/motion_agl1.py:
```diff
@@ -0,0 +1,194 @@
+    .friendly_name(model="Presence Sensor FP1E", manufacturer="Aqara")
+    .adds(DeviceTemperature)
+    .adds(OccupancySensingLocal)
+    .adds(
+        IasZoneLocal,
+    )
```
```suggestion
    .adds(IasZoneLocal)
```

TheJulianJES committed:
Format adding `IasZoneLocal`

TheJulianJES committed:
Add test for Aqara FP1E sensor

TheJulianJES committed:
Rearrange values in test

TheJulianJES commented:
I've also added a small test to cover the `_update_attribute` method: https://github.com/zigpy/zha-device-handlers/pull/3521/commits/f0d075478de26c6200023ec4a1971a3e9e1b984c

TheJulianJES reviewed (approved):
This should be good to go IMO. Thanks, @jeverley!

jeverley commented:
> I've also added a small test to cover the `_update_attribute` method: [f0d0754](https://github.com/zigpy/zha-device-handlers/commit/f0d075478de26c6200023ec4a1971a3e9e1b984c)

Thanks for adding these, I was still wrapping my head around how best to introduce a v2 quirk amongst the other xiaomi motion devices.
I think we need to make a small tweak to the expected statuses for motion in here, possible values being:

```
class AqaraMotion(types.enum8):
    """Aqara motion attribute values."""

    Idle = 0x02 # no presence
    Moving = 0x03 # present and moving
    Still = 0x04 # present and not moving
```
Therefore the parameters should be (if I'm understanding the test syntax correctly):

```
@pytest.mark.parametrize(
    "occupancy_value, expected_occ_status, motion_value, expected_motion_status",
    [
        (0, OccupancySensing.Occupancy.Unoccupied, 2, 0),
        (1, OccupancySensing.Occupancy.Occupied, 3, IasZone.ZoneStatus.Alarm_1),
        (1, OccupancySensing.Occupancy.Occupied, 4, 0),
    ],
)

```
Will add a commit reflecting this shortly.

Edit: I see you already corrected the flipped occupied/unoccupied values in https://github.com/zigpy/zha-device-handlers/pull/3521/commits/45c49fc50091eabba0981387ca104624cb8e9be5, still good to capture the 3rd scenario of occupied with no movement.

jeverley committed:
Add test scenario for presence with no motion

jeverley committed:
Merge branch 'dev' into lumi.sensor_occupy.agl1

TheJulianJES commented on zhaquirks/xiaomi/aqara/motion_agl1.py:
```diff
@@ -0,0 +1,192 @@
+
+(
+    QuirkBuilder("aqara", "lumi.sensor_occupy.agl1")
+    .friendly_name(model="Presence Sensor FP1E", manufacturer="Aqara")
```
I know it's `model, manufacturer` in zigpy, but let's do it the other way around to match the order of `QuirkBuilder`.
Since they are keyword arguments here, the order technically doesn't matter. We should still consider swapping it around in zigpy to align with the typical `manufacturer, model` order though.
```suggestion
    .friendly_name(manufacturer="Aqara", model="Presence Sensor FP1E")
```

TheJulianJES committed:
Swap around order for `friendly_name`

TheJulianJES commented on zhaquirks/xiaomi/aqara/motion_agl1.py:
```diff
@@ -0,0 +1,192 @@
+    Occupied = 0x01
+
+
+class IasZoneLocal(LocalDataCluster, IasZone):
+    """Virtual cluster for IasZone."""
+
+    _CONSTANT_ATTRIBUTES = {
+        IasZone.AttributeDefs.zone_type.id: IasZone.ZoneType.Motion_Sensor
+    }
+    _VALID_ATTRIBUTES = {IasZone.AttributeDefs.zone_status.id}
```
Note for other reviewers: We need a virtual cluster for `IasZone` here, as we need a virtual attribute for "motion". The Xiaomi `motion` attribute needs to be parsed in a way that we can't yet do with quirks v2.

Also, for motion/occupancy, it makes sense to use the integrated entities in ZHA IMO. We'd have to duplicate "motion" and "occupancy" quirk entities with all properties everywhere otherwise.

We already use virtual clusters in a lot of quirks and they'll always be needed for some (especially for Tuya and some Xiaomi devices), unless we have some kind of other structure we can map attributes to (like Z2M does).

TheJulianJES commented:
Looks like Codecov is broken across all repos again. This PR is fully covered.
Also noting that this quirk will require the latest ZHA version. You can't install it on a HA Core 2024.11.x system (or earlier) anymore.
